lib/mapObjects/CGHeroInstance.cpp

```diff
@@ -1893,7 +1893,7 @@ bool CGHeroInstance::isMissionCritical() const
 
 void CGHeroInstance::fillUpgradeInfo(UpgradeInfo & info, const CStackInstance & stack) const
 {
-	TConstBonusListPtr lista = getBonusesOfType(BonusType::SPECIAL_UPGRADE, BonusSubtypeID(stack.getId()));
+	TConstBonusListPtr lista = getBonuses(Selector::typeSubtype(BonusType::SPECIAL_UPGRADE, BonusSubtypeID(stack.getId())));
```

Comment:
@IvanSavenko Has to revert your change from cached bonuses from #5110 as it breaks the upgrade buttons.

Reply:
This is incorrect fix.
Problem is in getBonusesOfType method. Correct fix is
```diff
diff --git a/lib/bonuses/IBonusBearer.cpp b/lib/bonuses/IBonusBearer.cpp
index fa3f8b250..2e6f29d12 100644
--- a/lib/bonuses/IBonusBearer.cpp
+++ b/lib/bonuses/IBonusBearer.cpp
@@ -59,7 +59,7 @@ TConstBonusListPtr IBonusBearer::getBonusesOfType(BonusType type) const
 TConstBonusListPtr IBonusBearer::getBonusesOfType(BonusType type, BonusSubtypeID subtype) const
 {
        std::string cachingStr = "type_" + std::to_string(static_cast<int>(type)) + "_" + subtype.toString();
-       CSelector s = Selector::type()(type);
+       CSelector s = Selector::typeSubtype(type, subtype);
        return getBonuses(s, cachingStr);
 }
```